streams/src/main/java/org/apache/kafka/streams/processor/internals/KafkaStreamsInstances.java

```diff
@@ -0,0 +1,218 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals;
+
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.serialization.Serializer;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.processor.StreamPartitioner;
+import org.apache.kafka.streams.processor.TopologyBuilder;
+import org.apache.kafka.streams.state.HostInfo;
+import org.apache.kafka.streams.state.KafkaStreamsInstance;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.apache.kafka.common.utils.Utils.toPositive;
+
+/**
+ * Provides access to the {@link KafkaStreamsInstance} in a KafkaStreams application. This can be used
+ * to discover the locations of {@link org.apache.kafka.streams.processor.StateStore}s
+ * in a KafkaStreams application
+ */
+public class KafkaStreamsInstances {
+    private final Map<HostInfo, Set<TopicPartition>> hostToTopicPartition;
+    private final TopologyBuilder builder;
+    private final Map<String, List<TopicPartition>> partitionsByTopic;
+
+    public KafkaStreamsInstances(final Map<HostInfo, Set<TopicPartition>> hostToTopicPartition,
+                                 final TopologyBuilder builder) {
+        this.hostToTopicPartition = hostToTopicPartition;
+        this.partitionsByTopic = getPartitionsByTopic();
+        this.builder = builder;
+    }
+
+    /**
+     * Find all of the {@link KafkaStreamsInstance}s in a
+     * {@link KafkaStreams application}
+     *
+     * @return all the {@link KafkaStreamsInstance}s in a {@link KafkaStreams} application
+     */
+    public Collection<KafkaStreamsInstance> getAllStreamsInstances() {
+        final List<KafkaStreamsInstance> all = new ArrayList<>();
```

Comment:
Should we cache this result and invalidate if `hostToTopicPartition` gets updated?